### DISCUSSION TITLE
Can't make changes to `db_default`

### DISCUSSION THREAD

**Discussion by mikestreety** [sourcebroker/deployer-extended-typo3#28] [General]
Yo!

I was wanting to make some changes to the `db_default` variable (add some SQL to `post_sql_in`) which is set in each version's [set.php](https://github.com/sourcebroker/deployer-extended-typo3/blob/master/deployer/11/set.php#L36).

_**Sidenote:** I realise this code is actually called in [deployer-extended-typo3](https://github.com/sourcebroker/deployer-extended-typo3), however it is the implementation in this package which is causing the issue_

It seems because the variable is then used inside the [`db_databases`](https://github.com/sourcebroker/deployer-extended-typo3/blob/master/deployer/11/set.php#L67), it seems it gets cached and you cannot override without then overriding `db_databases`.

My current solution is the following: (I have managed to reduce `db_databases` as I want to use the TYPO3 CMS Driver)

```php
set('db_default', array_merge(get('db_default'), [
    'post_sql_in' => '[SQL STATEMENT HERE]',
]));
set('db_databases', [
	'database_default' => [
		get('db_default'),
		function () {
			return (new \SourceBroker\DeployerExtendedTypo3\Drivers\Typo3CmsDriver)->getDatabaseConfig();
		}
	]
]);
```

Is there anyway the script can re-evaluate at runtime or similar (or can we get a second, empty array which gets array merged) later on inside the `extended-db` package?

Happy to create an MR where it is needed, but would appreciate guidance on how and where

**Comment by kszymukowicz** [comment]
Unfortunately the current solution with `db_default` can not be refreshed automatically so in fact `db_default` is only a kind of internal var. When we overwrite db config on our projects we overwrite `db_databases` directly.

The only way to fix it (allow `db_default`) would be to declare `db_databases` as a closure. This way the "run" for this variable is delayed to first get() and you would have the possibility to overwrite `db_default` in deploy.php. This however needs a small change on the `deployer-extended-database` package and would be a big breaking change for all who extended `db_databases` treating it as array only.

It would be nice to have it improved but I am afraid the side effects overcome the gains. So for now the way you do it now - means redeclare `db_databases` - is the preferred way.

**Comment by mikestreety** [comment]
Hey @kszymukowicz Sorry to bring this up again.

I was trying to add some tables to the `ignore_tables_out` command but it doesn't seem to work. I use `deployer-extended-typo3` on top of this, so that might be related.

I have the following in my `deploy.php`

```php
set(
	'db_default',
	[
		// Ignore tables
		'ignore_tables_out' => [
			'cf_.*',
			'cache_.*',
			'be_sessions',
			'fe_sessions',
			'fe_session_data',
			'sys_file_processedfile',
			'sys_history',
			'sys_log',
			'sys_refindex',

			// Custom additions
			'tx_kesearch_index',
			'tx_simplelog404_domain_model_logentry',
		],
		// Delete non-liquid light users & remove MFA
		'post_sql_in' => '
			UPDATE be_users SET mfa = "";
		',
	]
);

// Needs to be redeclared to reset the variable from above
set('db_databases', [
	'database_default' => [
		get('db_default'),
		function () {
			return (new \SourceBroker\DeployerExtendedTypo3\Drivers\Typo3CmsDriver())->getDatabaseConfig();
		},
	],
]);
```

Despite this, `tx_kesearch_index` is still full up after a pull. Am I doing something wrong with the config?

Thanks

**Comment by kszymukowicz** [comment]
hi @mikestreety 

what I am doing is

```
set('db_databases_overwrite_global', [
    'database_default' =>
        [
            '0' =>
                [
                    'ignore_tables_out' =>
                        [
                            0 => 'tx_myext_model',
                            1 => 'tx_myext_model',
                        ],
                ],
        ],
]);
```

**Comment by kszymukowicz** [comment]
Maybe you switched from 19 to 20?
https://github.com/sourcebroker/deployer-extended-typo3/blob/master/UPGRADE.rst

**Reply by mikestreety** [reply]
Ah! I missed that - thanks :)

**Reply by mikestreety** [reply]
Hmm, this still doesn't seem to work :( i'm on 

- deployer v7.4.1
- deployer-extended-typo3 21.0.0 and 
- deployer-extended-database 17.0.0

```php
set('db_databases_overwrite_global', [
	'database_default' => [
		[
			// Ignore tables
			'ignore_tables_out' => [
				'cf_.*',
				'cache_.*',
				'be_sessions',
				'fe_sessions',
				'fe_session_data',
				'sys_file_processedfile',
				'sys_history',
				'sys_log',
				'sys_refindex',
				'tx_kesearch_index',
				'tx_simplelog404_domain_model_logentry',
			],
			// Delete non-liquid light users & remove MFA
			'post_sql_in' => '
				UPDATE be_users SET mfa = "";
			',
		],
		function () {
			return (new \SourceBroker\DeployerExtendedTypo3\Drivers\Typo3CmsDriver())->getDatabaseConfig();
		},
	],
]);
```

**Reply by mikestreety** [reply]
To add a bit more, it seems to be executing the `post_sql_in` but not ignoring the `tx_kesearch_index` table

**Reply by kszymukowicz** [reply]
and what you see when you do
```
print_r(get('db_databases_merged'));
die;
```
under `set('db_databases_overwrite_global',`

**Reply by kszymukowicz** [reply]
try to remove
```
		function () {
			return (new \SourceBroker\DeployerExtendedTypo3\Drivers\Typo3CmsDriver())->getDatabaseConfig();
		},		
```

**Reply by mikestreety** [reply]
Ah, removing the `function()` seems to have done it.

Thanks!
